src/client/app/utils/determineCompatibleUnits.ts

```diff
@@ -0,0 +1,146 @@
+			compatibleUnits = intersect(compatibleUnits, meterUnits);
+		}
+	});
+	// Now have final compatible units for the provided set of meter
+	return compatibleUnits;
+}
+
+/**
+ * Returns a set of units ids that are compatible with a specific unit id.
+ * @param unit The unit id.
+ * @returns
+ */
+export function unitsCompatibleWithUnit(unitId: number): Set<number> {
+	// unitSet starts as an empty set.
+	const unitSet = new Set<number>();
+	// If unit was null in the database then -99. This means there is no unit
+	// so nothing is compatible with it. Skip processing and return empty set at end.
+	if (unitId != -99) {
+		// The Pik array.
+		const pik = ConversionArray.pik;
+		// Get the row index in Pik of this unit.
+		const row = pRowFromUnit(unitId);
+		// The compatible units are all columns with true for Pik where i = row.
+		// Loops over all columns of Pik in row.
+		for (let k = 0; k < pik[0].length; ++k) {
+			if (pik[row][k]) {
+				// unit at index k is compatible with meter unit so add to set.
+				// Convert index in Pik to unit id.
+				unitSet.add(unitFromPColumn(k));
+			}
+		}
+	}
+	return unitSet;
+}
+
+/**
+ * Returns the row index in Pik for a meter unit.
+ * @param unitId The unit id.
+ * @returns
+ */
+export function pRowFromUnit(unitId: number): number {
+	const state = store.getState();
+	const unit = _.find(state.units.units, function (o: UnitData) {
+		// Since this is the row index, type of unit must be meter.
+		return o.id == unitId && o.typeOfUnit == UnitType.meter;
+	}) as UnitData;
+	return unit.unitIndex;
+}
+
+/**
+ * Returns the unit id given the row in Pik.
+ * @param row The row to find the associated unit.
+ * @returns
+ */
+export function unitFromPRow(row: number): number {
+	const state = store.getState();
+	const unit = _.find(state.units.units, function (o: UnitData) {
+		// Since the given unitIndex is a row index, the unit type must be meter.
+		return o.unitIndex == row && o.typeOfUnit == UnitType.meter;
+	}) as UnitData;
+	return unit.id;
+}
+
+/**
+ * Returns the unit id given the column in Pik.
+ * @param column The column to find the associated unit.
+ * @returns
+ */
+export function unitFromPColumn(column: number): number {
+	const state = store.getState();
+	const unit = _.find(state.units.units, function (o: UnitData) {
+		// Since the given unitIndex is a column index, the unit type must be different from meter.
+		return o.unitIndex == column && o.typeOfUnit != UnitType.meter;
+	}) as UnitData;
+	return unit.id;
+}
+
+/**
+ * Returns the set of meters's ids associated with the groupId used.
+ * @param groupId The groupId.
+ * @returns
+ */
+export async function metersInGroup(groupId: number): Promise<Set<number>> {
+	// Fetch group children if needed.
+	await store.dispatch<any>(fetchGroupChildrenIfNeeded(groupId));
+
+	const state = store.getState();
+	// Gets the group associated with groupId.
+	const group = _.find(state.groups.byGroupID, function (o: UnitData) {
```

Comment:
I tried to do so but got the following error:

```Element implicitly has an 'any' type because expression of type 'number' can't be used to index type 'Dictionary<NamedIDItem> | MeterMetadataByID'.```

The state that we use in other places has the type of State, while it's an object here. This may be the reason. I have done some google searches but still couldn't find a way to make it work. Please let me know if you have a way to work around this.

Alternatively, I tried: '_.get(state.meters.byMeterID, meterId)' and it works properly. Please let me know what you think about this.

Reply:
Okay. There may be a way to modify the type definition but I don't feel it is important enough to do it. Your alternative seems a little more straightforward so why not use that (but the original way is okay too).